Fix memory safety of  `Array.copy_to` (#4173)
ponylang/ponyc: .release-notes/4173.md

## Fix for segfault when using Array.copy_to() with an empty Array

> Currently, the `Array.copy_to()` function does not check whether the Array is empty or has been instanced with any values. This leads to a runtime crash when calling `copy_to()` on an empty array such as in this example (issue #4172):

Reviewer: You can't say currently here. Release notes are after the change it made. This would be...

> Currently, the `Array.copy_to()` function does not check whether the Array is empty or has been instanced with any values. This leads to a runtime crash when calling `copy_to()` on an empty array such as in this example (issue #4172):

Reviewer: ```suggestion
Previously, `Array.copy_to()` did not check whether the source Array had been initialized. Uninitialized arrays internally have a null pointer for their data as no memory has been allocated for them yet. When doing a copy like in the example below, this would result in an attempt to dereference the null pointer and a segfault.
```

```pony
class Foo

>     embed _array: Array[I32] = []

Reviewer: ```suggestion
  embed _array: Array[I32] = []
```

> new create() => None

Reviewer: ```suggestion
  new create() => 
    None

```

> new clone(src: Foo) => src._array.copy_to(_array, 0, 0, 10)

Reviewer: ```suggestion
  new clone(src: Foo) =>
    src._array.copy_to(_array, 0, 0, 10)
```

actor Main

> 

Reviewer: ```suggestion
```

>     new create(env': Env) =>

Reviewer: ```suggestion
  new create(env': Env) =>
```

>         let f2 = Foo.clone(Foo)

Reviewer: ```suggestion
  let foo1 = Foo
  let foo2 = Foo.clone(foo1)
```

```

> 

Reviewer: ```suggestion

`Array` is part of the `builtin` package of the standard library. Only code in `builtin` is allowed to do direct pointer manipulations like those that `Array` uses to implement `copy_to`. Direct pointer manipulations, if done incorrectly, can lead to segfaults and other memory related errors. Because pointer manipulation can be dangerous, only code in the `builtin` package is allowed to do it. 

We consider unsafe pointer operations in the `builtin` package to be of utmost importance and once this bug was found moved quickly to release a fix.
```

> This PR is a preliminary fix that safeguards for such cases -- calling `copy_to()` on an empty Array will simply do nothing.

Reviewer: ```suggestion
```
